File: java/app/src/main/java/com/java/aoc23/Day1.java
/*
 * Advent of Code 2023 Day 1
 */
package com.java.aoc23;

import java.util.ArrayList;

public class Day1 implements Day {
    public void part1(String input) {
        String[] calibration_values = input.split("\n");
        int sum = 0;
        for (String calibration_value : calibration_values) {
            ArrayList<String> numbers = new ArrayList<String>();
            for (int i = 0; i < calibration_value.length(); i++) {
                if (Character.isDigit(calibration_value.charAt(i))) {
                    numbers.add(calibration_value.substring(i, i + 1));
                }
            }
            int number = Integer.parseInt(numbers.get(0)) * 10 + Integer.parseInt(numbers.get(numbers.size() - 1));
            sum += number;
        }
        System.out.println("Part 1: " + sum);
    }

    public void part2(String input) {
        System.out.println("Part 2 not implemented yet");
    }
}

File: rust/src/day1.rs
/*
 * Advent of Code 2023 Day 1
 */
use fancy_regex::Regex;

pub struct Day1;

impl Day1 {
    pub fn part1(&self, input: String) {
        let calibration_values: Vec<&str> = input.split("\n").collect();
        let mut sum: i32 = 0;
        for value in calibration_values {
            let mut numbers: Vec<i32> = Vec::new();
            for character in value.chars() {
                if character.is_digit(10) {
                    numbers.push(character.to_digit(10).unwrap() as i32);
                }
            }
            let number: i32 = numbers[0] * 10 + numbers.last().unwrap();
            sum += number;
        }
        println!("Part 1: {}", sum);     
    }

    pub fn part2(&self, input: String) {
        let calibration_values: Vec<&str> = input.split("\n").collect();
        let mut sum: i32 = 0;
        for value in calibration_values {
            let numbers = self.parse_digits(value.to_string());
            let number: i32 = numbers[0] * 10 + numbers.last().unwrap();
            sum += number;
        }
        println!("Part 2: {}", sum);
    }

    pub fn parse_digits(&self, input: String) -> Vec<i32> {
        let re = Regex::new(r"(?=(one|two|three|four|five|six|seven|eight|nine|\d))").unwrap();
        let number_strings: Vec<&str> = re.captures_iter(&input).map(|mat| mat.unwrap().get(1).unwrap().as_str()).collect();
        let mut numbers: Vec<i32> = Vec::new();
        for number_string in number_strings {
            numbers.push(match number_string {
                "one" => 1,
                "two" => 2,
                "three" => 3,
                "four" => 4,
                "five" => 5,
                "six" => 6,
                "seven" => 7,
                "eight" => 8,
                "nine" => 9,
                _ => number_string.parse::<i32>().unwrap(),
            });
        }
        numbers
    }
}